kart/annotations/__init__.py

```diff
@@ -0,0 +1,70 @@
+import json
+import logging
+from .db import annotations_session, KartAnnotation
+
+L = logging.getLogger(__name__)
+
+
+class DiffAnnotations:
+    def __init__(self, repo):
+        self.repo = repo
+
+    def _object_id(self, base_rs, target_rs):
+        # this is actually symmetric, so we can marginally increase hit rate by sorting first
+        if base_rs.tree.id < target_rs.tree.id:
+            return f"{base_rs.tree.id}...{target_rs.tree.id}"
+        else:
+            return f"{target_rs.tree.id}...{base_rs.tree.id}"
+
+    def store(self, *, base_rs, target_rs, annotation_type, data):
+        """
+        Stores a diff annotation to the repo's sqlite database,
+        and returns the annotation itself.
+        """
+        assert isinstance(data, dict)
+        with annotations_session(self.repo) as session:
+            object_id = self._object_id(base_rs, target_rs)
+            data = json.dumps(data)
+            L.debug(
+                "storing: %s for %s: %s",
+                annotation_type,
+                object_id,
+                data,
+            )
+            session.add(
```

Comment:
Needs to upsert for concurrency... or handle that via catching the `COMMIT` error and ignoring?

Reply:
And read-only DBs/directories will obviously fail to write

Reply:
> upsert

this is only necessary if there is a unique constraint on `(object_type, object_id, annotation_type)`.

It's easiest to just not have that constraint. Then there's a possibility of more than one transaction inserting the same annotation, but we can just allow that and not have to worry about concurrency (in `get` I just fetch the first one, and don't worry too hard about the presence of others)

Reply:
A unique constraint can have a default/automatic `ON CONFLICT` resolution, _and_ it's supported by SQLAlchemy... https://docs.sqlalchemy.org/en/14/dialects/sqlite.html#on-conflict-support-for-constraints

You're presuming `data` is immutable as well?

Reply:
yes data is immutable if we always use object IDs in the `object_id` field, and not refs or whatever

Reply:
well, it's not really - if the counting method changes, we record different things (eg: adding add/mod/delete counts with some sort of "renames detection happened" flag) then `data` will change. Or even JSON not serialising deterministically (though iirc it's sorted in py3?)

Reply:
ok, done

table now looks like 
```sql
CREATE TABLE kart_annotations (
	id INTEGER NOT NULL,
	object_id TEXT NOT NULL,
	annotation_type TEXT NOT NULL,
	data TEXT NOT NULL,
	PRIMARY KEY (id),
	CONSTRAINT kart_annotations_multicol UNIQUE (annotation_type, object_id) ON CONFLICT REPLACE
)
```